app/controllers/mailing_list_signups_controller.rb

```diff
@@ -0,0 +1,89 @@
+class MailingListSignupsController < ApplicationController
+  include CircuitBreaker
+  layout "registration"
+
+  SESSION_STORE_KEY = :mailing_list_signup
+
+  def new
+    @signup = MailingList::Signup.new
+  end
+
+  def create
+    @signup = MailingList::Signup.new(new_signup_params)
+
+    if @signup.valid?
+      save_signup_to_session
+
+      return redirect_to(edit_mailing_list_signup_path) if @signup.already_signed_up?
+
+      @signup.add_member_to_mailing_list!
+      complete
+    else
+      render :new
+    end
+  end
+
+  def edit
+    restore_signup_from_session
+  end
+
+  def update
+    restore_signup_from_session
+
+    @signup.assign_attributes(edit_signup_params)
+
+    if @signup.valid?(:verify)
+      save_signup_to_session
+
+      return complete("already_subscribed") if @signup.already_subscribed_to_mailing_list
+
+      @signup.add_member_to_mailing_list!
+      complete
+    else
+      render :edit
+    end
+  end
+
+private
+
+  def complete(path = "completed")
+    session.delete(SESSION_STORE_KEY)
+    redirect_to mailing_list_step_path(path)
+  end
+
+  def restore_signup_from_session
+    signup_data = session[SESSION_STORE_KEY]
+
+    if signup_data.nil?
+      Rails.logger.warn("tried to confirm identity with blank session")
+      return redirect_to(new_mailing_list_signup_path)
```

Comment:
Could this not result in odd bugs if we return a `redirect_to` the caller will continue to process the code after the call to `restore_signup_from_session` (so line 33 onward in this file)?

Reply:
Yep agree 🤦🏽 

Reworked with a VerifyWithEmptySession error